### PR TITLE
Implemented AlphaVantageDataDownloader

### PR DESCRIPTION
**Pull request by jdharmon**
#### Description
Implemented `IDataDownloader` for Alpha Vantage. Supports equities in daily, hourly, and minute resolutions.

#### Related Issue
Requested in https://www.quantconnect.com/forum/discussion/9283/alphavantage-data-for-backtesting-locally-in-visual-studio/p1

#### Motivation and Context
Adds support for downloading local data from free Alpha Vantage API.

#### Requires Documentation Change
No.

#### How Has This Been Tested?
Tested data download in all resolutions. Full-coverage unit tests included.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [X] I have added tests to cover my changes. 
- [X] All new and existing tests passed.
- [X] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
+﻿using CsvHelper;
```
Comment (jaredbroad) [inline ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs:1]: Please add the LEAN Apache2.0 header to all files -- (can see on other code files)

File: ToolBox/AlphaVantageDownloader/AlphaVantageAuthenticator.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace QuantConnect.ToolBox.AlphaVantageDownloader
+{
+    class AlphaVantageAuthenticator : IAuthenticator
+    {
+        private readonly string _apiKey;
+
+        public AlphaVantageAuthenticator(string apiKey)
```
Comment (jaredbroad) [inline ToolBox/AlphaVantageDownloader/AlphaVantageAuthenticator.cs:15]: Please XML comment all public classes and methods even for baby classes

File: Tests/ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloaderTests.cs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace QuantConnect.Tests.ToolBox.AlphaVantageDownloader
+{
+    [TestFixture]
+#pragma warning disable CA1001 // Types that own disposable fields should be disposable
+    public class AlphaVantageDataDownloaderTests
```
Comment (jaredbroad) [inline Tests/ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloaderTests.cs:17]: ❤️  👏 Nice tests!

File: ToolBox/AlphaVantageDownloader/AlphaVantageDownloaderProgram.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+                // fix end date
+                endDate = new DateTime(Math.Min(endDate.Ticks, DateTime.Now.AddDays(-1).Ticks));
+
+                // Load settings from config.json
+                var dataDirectory = Config.Get("data-folder", "../../../Data");
+
+                // Only Equity for now
+                SecurityType securityType = SecurityType.Equity;
+                string market = Market.USA;
```
Comment (jaredbroad) [inline ToolBox/AlphaVantageDownloader/AlphaVantageDownloaderProgram.cs:40]: think you can inline this, used once

File: ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        /// Execute request and parse response.
+        /// </summary>
+        /// <param name="request">The request</param>
+        /// <returns><see cref="TimeSeries"/> data</returns>
+        private IEnumerable<TimeSeries> GetTimeSeries(RestRequest request)
+        {
+            var response = _avClient.Get(request);
+
+            using (var reader = new StringReader(response.Content))
+            using (var csv = new CsvReader(reader, CultureInfo.InvariantCulture))
```
Comment (jaredbroad) [inline ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs:126]: Please wrap in braces we try and avoid implicitly bracketed code

File: ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        /// <param name="request">Base request</param>
+        /// <param name="startUtc">Start time</param>
+        /// <param name="endUtc">End time</param>
+        /// <returns></returns>
+        private IEnumerable<TimeSeries> GetDailyData(RestRequest request, DateTime startUtc, DateTime endUtc)
+        {
+            request.AddParameter("function", "TIME_SERIES_DAILY");
+            // The default output only includes 100 trading days of data. If we want need more, specify full output
+            if (GetBusinessDays(startUtc, endUtc) > 100)
+                request.AddParameter("outputsize", "full");
```
Comment (jaredbroad) [inline ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs:77]: Brackets

**Review by jaredbroad — CHANGES_REQUESTED** [review]
Overall solid PR @jdharmon! Some style and comment tweaks to match LEAN but should be quick fix.

File: ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+                yield return $"year{year}month{month}";
+            }
+        }
+
+        /// <summary>
+        /// From https://stackoverflow.com/questions/1617049/calculate-the-number-of-business-days-between-two-dates
+        /// </summary>
+        public static double GetBusinessDays(DateTime start, DateTime end)
+        {
+            double days = ((end - start).TotalDays * 5 - (start.DayOfWeek - end.DayOfWeek) * 2) / 7;
```
[the remark this answers is not in the document]
Comment (jdharmon) [reply ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs:160]: I knew this function had to exist somewhere, but I could not find it.

Unfortunately, `Time.TradeableDates()` takes a `Security` argument. Any guidance on how to create a `Security` outside of an algorithm?

File: ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+                    break;
+                case Resolution.Daily:
+                    data = GetDailyData(request, startUtc, endUtc);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(resolution), $"{resolution} resolution not supported by API.");
+            }
+
+            var period = resolution.ToTimeSpan();
+            return data.Select(d => new TradeBar(d.Time, symbol, d.Open, d.High, d.Low, d.Close, d.Volume, period));
```
Comment (nuaimat) [inline ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs:84]: Hi,
I cloned this pull request on locally, I am running on a Mac OS using mono but I am seeing the following issue when executing ```mono QuantConnect.ToolBox.exe 
    --app=alphavantagedownloader 
    --tickers=SPXL 
    --resolution=Hour 
    --from-date=20210101-00:00:00 \
    --api-key=```
```20210423 11:48:52.933 ERROR:: AlphaVantageDownloaderProgram.AlphaVantageDownloader():  CsvHelper.HeaderValidationException: Header with name 'time' or 'timestamp'[0] was not found.
Header with name 'open'[0] was not found.
Header with name 'high'[0] was not found.
Header with name 'low'[0] was not found.
Header with name 'close'[0] was not found.
Header with name 'volume'[0] was not found.
If you are expecting some headers to be missing and want to ignore this validation, set the configuration HeaderValidated to null. You can also change the functionality to do something else, like logging the issue.

  at CsvHelper.Configuration.ConfigurationFunctions.HeaderValidated (CsvHelper.InvalidHeader[] invalidHeaders, CsvHelper.ReadingContext context) [0x0006d] in <e0ff1191374f4bc2b71069db98504b37>:0
  at CsvHelper.CsvReader.ValidateHeader (System.Type type) [0x000af] in <e0ff1191374f4bc2b71069db98504b37>:0
  at CsvHelper.CsvReader.ValidateHeader[T] () [0x00000] in <e0ff1191374f4bc2b71069db98504b37>:0
  at CsvHelper.CsvReader+<GetRecords>d__63`1[T].MoveNext () [0x0006f] in <e0ff1191374f4bc2b71069db98504b37>:0
  at System.Collections.Generic.EnumerableHelpers.ToArray[T] (System.Collections.Generic.IEnumerable`1[T] source, System.Int32& length) [0x00032] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at System.Linq.Buffer`1[TElement]..ctor (System.Collections.Generic.IEnumerable`1[T] source) [0x00022] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at System.Linq.OrderedEnumerable`1[TElement].ToList () [0x00000] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at QuantConnect.ToolBox.AlphaVantageDownloader.AlphaVantageDataDownloader.GetTimeSeries (RestSharp.RestRequest request) [0x00052] in <8e97fb96193a498bad25e020041ad1a8>:0
  at QuantConnect.ToolBox.AlphaVantageDownloader.AlphaVantageDataDownloader+<GetIntradayData>d__5.MoveNext () [0x00136] in <8e97fb96193a498bad25e020041ad1a8>:0
  at System.Linq.Enumerable+SelectEnumerableIterator`2[TSource,TResult].MoveNext () [0x00029] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] keySelector, System.Func`2[T,TResult] elementSelector, System.Collections.Generic.IEqualityComparer`1[T] comparer) [0x000a3] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] keySelector, System.Func`2[T,TResult] elementSelector) [0x00000] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at QuantConnect.ToolBox.LeanDataWriter.WriteDailyOrHour (System.Collections.Generic.IEnumerable`1[T] source) [0x0001d] in <8e97fb96193a498bad25e020041ad1a8>:0
  at QuantConnect.ToolBox.LeanDataWriter.Write (System.Collections.Generic.IEnumerable`1[T] source) [0x00018] in <8e97fb96193a498bad25e020041ad1a8>:0
  at QuantConnect.ToolBox.AlphaVantageDownloader.AlphaVantageDownloaderProgram.AlphaVantageDownloader (System.Collections.Generic.List`1[T] tickers, System.String resolution, System.DateTime fromDate, System.DateTime toDate, System.String apiKey) [0x00100] in <8e97fb96193a498bad25e020041ad1a8>:0```

any ideas how to fix it?

Comment (jdharmon) [reply ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs:84]: @nuaimat It looks like you might be getting rate limited. You can try with a shorter `from-date` and `to-date` window.

Unfortunately, the API always responds with a `200 OK` regardless of status. I'll add some extra logic to look for content other than CSV data and display the message to the user.

Comment (nuaimat) [reply ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs:84]: @jdharmon that's correct, printing the values inside GetTimeSeries shows the rate limit message from the api,  
adding `System.Threading.Thread.Sleep(15*1000);` inside GetIntradayData `foreach (var slice in slices)`
seems to do the trick for me, would love if this was parametrized "--delayBetweenRequests"

Comment (jdharmon) [reply ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs:90]: Added rate limiting.

**Review by Martin-Molinero — APPROVED** [review]
Thanks! Awesome PR, push minor tweaks and rebased ✅

**Pull request merged**